crypto/src/main/java/org/springframework/security/crypto/bcrypt/BCryptPasswordEncoder.java

```diff
@@ -51,32 +53,77 @@ public BCryptPasswordEncoder(int strength) {
 		this(strength, null);
 	}
 
+	/**
+	 * @param version the version of bcrypt, can be 2a,2b,2y
+	 */
+	public BCryptPasswordEncoder(String version) {
+		this(version, null);
+	}
+
+	/**
+	 * @param version the version of bcrypt, can be 2a,2b,2y
+	 * @param random the secure random instance to use
+	 */
+	public BCryptPasswordEncoder(String version, SecureRandom random) {
+		this(version, -1, random);
+	}
+
 	/**
 	 * @param strength the log rounds to use, between 4 and 31
 	 * @param random the secure random instance to use
 	 *
 	 */
 	public BCryptPasswordEncoder(int strength, SecureRandom random) {
+		this(null, strength, random);
+	}
+
+	/**
+	 * @param version the version of bcrypt, can be 2a,2b,2y
+	 * @param strength the log rounds to use, between 4 and 31
+	 */
+	public BCryptPasswordEncoder(String version, int strength) {
+		this(version, strength, null);
+	}
+
+	/**
+	 * @param version the version of bcrypt, can be 2a,2b,2y
+	 * @param strength the log rounds to use, between 4 and 31
+	 * @param random the secure random instance to use
+	 *
+	 */
+	public BCryptPasswordEncoder(String version, int strength, SecureRandom random) {
 		if (strength != -1 && (strength < BCrypt.MIN_LOG_ROUNDS || strength > BCrypt.MAX_LOG_ROUNDS)) {
 			throw new IllegalArgumentException("Bad strength");
 		}
+		this.version = version;
 		this.strength = strength;
 		this.random = random;
 	}
 
 	public String encode(CharSequence rawPassword) {
 		String salt;
-		if (strength > 0) {
-			if (random != null) {
-				salt = BCrypt.gensalt(strength, random);
+		if(version != null) {
```

Comment:
Rather than having this check and two different code paths, let's make it so that version is final and always non-null with a default of $2y